### PR TITLE
[DOCS] Added New doc page for Flyte Type system and Data Passing

### PR DESCRIPTION
**Pull request by 10sharmashivam**
## Tracking issue
Reference #3322 

## Why are the changes needed?

This documentation aims to provide users with a comprehensive understanding of Flyte’s type system and data passing mechanisms, ensuring they can effectively leverage these features in their workflows.

## What changes were proposed in this pull request?

	•	Created a new documentation page detailing the Flyte type system and data passing.
	•	Included code examples for primitive types, collections, and custom types to illustrate the concepts clearly.

## Additional Comments-

I intend to submit a follow-up PR in flyteorg/flytesnacks to include additional code examples under flyteorg/flytesnacks/examples/basics/basics. Once that PR is reviewed and merged, we can use the raw GitHub user content link for the examples and incorporate them into the documentation page, following the format used in other files in the documentation.

## Check all the applicable boxes 

- [ X] I updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ X] All commits are signed-off.

### CODE REVIEW

**Review by Future-Outlier — COMMENTED** [review]
Hi, can you follow the `Docs link` section like this PR?

https://github.com/flyteorg/flyte/pull/5844

File: docs/user_guide/basics/flyte_type_system.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+		```{rli} 
+			@task
+			def process_file(file: FlyteFile) -> str:
+    		# Simulate processing the file and returning its name
+    		return f"Processed file: {file}"
+		:caption: basics/task.py
+		:lines: 1
+		```
+
+### When to Use the Type System
```
Comment (davidmirror-ops) [inline docs/user_guide/basics/flyte_type_system.md:67]: Not sure if this section is needed. Sounds like there are situations when you don't need to use Flyte's type system and that's not the case. This is a core construct in Flyte

File: docs/user_guide/basics/flyte_type_system.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+### A Clear Mental Model
+
+At a high level, data passing between Flyte tasks and workflows can be thought of as a pipeline. Each task produces outputs, which are passed as inputs to the next task in the sequence. However, Flyte handles data passing in a way that ensures separation between the control plane (which manages workflow execution) and the data plane (where raw data is stored).
+
+There are two primary types of data in Flyte:
+
+	1.	Metadata: Information about the data, such as the structure of the data (e.g., data types, paths to data storage).
+	2.	Raw Data: The actual data itself (e.g., a Pandas DataFrame, a file in cloud storage).
+
+### How Data is Handled in Workflow Execution
```
Comment (davidmirror-ops) [inline docs/user_guide/basics/flyte_type_system.md:103]: I think this section repeats some of what is already covered [here](https://docs.flyte.org/en/latest/user_guide/concepts/main_concepts/data_management.html#divedeep-data-management)

**Comment by davidmirror-ops** [comment]
@10sharmashivam thanks for volunteering, this one is not a small feat! Would you be open to have a call where we can chat more about Flyte's type system, hear your questions and then we can eventually update this PR?

**Comment by 10sharmashivam** [comment]
> @10sharmashivam thanks for volunteering, this one is not a small feat! Would you be open to have a call where we can chat more about Flyte's type system, hear your questions and then we can eventually update this PR?

@davidmirror-ops I’d be happy to set up a call to discuss the Flyte-type system further and make sure the PR is aligned with Flyte’s core principles. I appreciate the opportunity to dive deeper into this, and I look forward to your insights!

**Comment by davidmirror-ops** [comment]
I think timezone difference will make it a bit hard (I'm on UTC-5). Tomorrow (Wednesday 23) 9:00 a.m. UTC-5 would work for you?

**Comment by 10sharmashivam** [comment]
Yes, that time works for me. Looking forward to the call.

**Comment by davidmirror-ops** [comment]
@10sharmashivam already synced with @eapolinario on this
1. Dynamic typing: if this is how we understand it (types that only exist at runtime) this is not accessible today in flytekit but you have to declare type hints upfront.
2. I think this section would be even better as a blog post on flyte.org. Would you be open to porting the content here to [this doc](https://docs.google.com/document/d/177ZWhP35F3qJgeBX4MwoAu42TUareWnxIydqhxneg7c/edit?usp=sharing)? There, we can iterate and add you as a co-author in the Flyte blog. wdyt?

**Comment by 10sharmashivam** [comment]
@davidmirror-ops Thank you for the opportunity! I would be thrilled to co-author the blog post for flyte.org

**Comment by 10sharmashivam** [comment]
Hi @davidmirror-ops, 

I have added content about dynamic typing & Static typing in flyte in the given [doc file](https://docs.google.com/document/d/177ZWhP35F3qJgeBX4MwoAu42TUareWnxIydqhxneg7c/edit?tab=t.0), starting from 3rd page.

I also applied feedbacks and suggestions giving earlier, and discussed in our google meet. Like reference to related already existing docs.

And I also submitted [PR](https://github.com/flyteorg/flytesnacks/pull/1761), about coding examples, which can be used (when approved, and merged) as raw github user content link, serving as code example snippets in the Flyte system documentation.

Thank you!

**Comment by 10sharmashivam** [comment]
Hi @davidmirror-ops,

I wanted to follow up about-
	1.	The content I added regarding dynamic and static typing in Flyte, in the shared document. Was it aligned with your expectations? If there are any changes needed or additional details you’d like, please let me know.
2. The Original Issue of adding a new page about flyte type system and data passing in documentation. Anything about that? the things we discussed in the meeting? Should I approach this issue with some new approach? or any changes in the current PRs that I submitted?

Thank you!

**Comment by davidmirror-ops** [comment]
@10sharmashivam as mentioned, the idea is to work in this as a blog post (with you as co author).
Ince that's posted we can close this issue by adding a reference in the docs to the blog.
Same with the example.

Wdyt?

**Comment by davidmirror-ops** [comment]
Sure! I'll keep adding content and comments to the shared doc

**Pull request closed without merging**